packages/augur-core/src/contracts/para/AMMExchange.sol

```diff
@@ -1,243 +1,325 @@
+
+        require(_poolInvalid > 0, "AugurCP: The pool doesn't have enough INVALID tokens to fulfill the request.");
+        require(_poolNo > 0, "AugurCP: The pool doesn't have enough NO tokens to fulfill the request.");
+        require(_poolYes > 0, "AugurCP: The pool doesn't have enough YES tokens to fulfill the request.");
+
+        // simulate user swapping YES to NO or NO to YES
+        uint256 _poolConstant = poolConstant(_poolYes, _poolNo);
+        if (_buyYes) {
+            // yesToUser + poolYes - poolConstant / (poolNo + _setsToBuy)
+            return _setsToBuy.add(_poolYes.sub(_poolConstant.div(_poolNo.add(_setsToBuy))));
+        } else {
+            return _setsToBuy.add(_poolNo.sub(_poolConstant.div(_poolYes.add(_setsToBuy))));
+        }
+    }
+
+    // If you do not have complete sets then you must have more shares than _setsToSell because you will be swapping
+    // some of them to build complete sets.
+    function exitPosition(uint256 _setsToSell, uint256 _maxSharesSwappedAway) public {
+        uint256 _cashToBuy = _setsToSell.mul(numTicks);
+        (uint256 _noFromUser, uint256 _yesFromUser) = rateExitPosition(_setsToSell);
+
+        uint256 _sharesSwappedAway = _setsToSell.mul(2).sub(_noFromUser).sub(_yesFromUser);
+        require(_sharesSwappedAway <= _maxSharesSwappedAway, "AugurCP: Could not exit position without trading away more shares than desired.");
+
+        // materialize the complete set sale for cash
+        shareTransfer(msg.sender, address(this), _setsToSell, _noFromUser, _yesFromUser);
+        cash.transfer(msg.sender, _cashToBuy);
+    }
+
+	function exitAll(uint256 _minCashPayout) external {
 		(uint256 _userInvalid, uint256 _userNo, uint256 _userYes) = shareBalances(msg.sender);
+		exitPositionShares(_userYes, _userNo, _userInvalid, _minCashPayout);
+	}
+
+	function exitPositionShares(uint256 _yesShares, uint256 _noShares, uint256 _invalidShares, uint256 _minCashPayout) public {
 		(uint256 _poolNo, uint256 _poolYes) = yesNoShareBalances(address(this));
-		uint256 _setsToSell = _cashToBuy.div(numTicks);
+		uint256 _invalidFromUser = _invalidShares;
+		uint256 _yesFromUser = _yesShares;
+		uint256 _noFromUser = _noShares;
+		uint256 _setsToSell = _invalidShares;
 
-		// short circuit if user is closing out their own complete sets
-		if (_userInvalid >= _setsToSell && _userNo >= _setsToSell && _userYes >= _setsToSell) {
-			shareTransfer(msg.sender, address(this), _setsToSell, _setsToSell, _setsToSell);
-			cash.transfer(msg.sender, _cashToBuy);
-			return;
+		// Figure out how many shares we're buying in our synthetic swap and use that to figure out the final balance of Yes/No (setsToSell)
+		if (_yesShares > _noShares) {
+			uint256 _delta = _yesShares.sub(_noShares);
+			uint256 _noSharesToBuy = quadratic(1, -int256(_delta.add(_poolYes).add(_poolNo)), int256(_delta.mul(_poolNo)), _yesShares);
```

Comment:
Seeing as how all call sites have the second parameter as negative, consider instead keeping these as positive `uint256` variables and just having the `quadratic` function subtract instead of add.  I believe this will actually clean up the quadratic formula quite a bit if you do a bit of algebra to extract out the sign.